bpf/tests/hostfw_extended_protocols.c

```diff
@@ -0,0 +1,336 @@
+#include "common.h"
+#include "pktgen.h"
+
+/* Enable code paths under test */
+#define ENABLE_IPV4			1
+#define ENABLE_HOST_FIREWALL		1
+
+#define NODE_IP				v4_node_one
+#define NODE_IP2			v4_node_two
+
+#define DST_IP			v4_ext_one
+#define DST_IP2			v4_ext_two
+
+static volatile const __u8 *node_mac = mac_one;
+static volatile const __u8 *dst_mac = mac_two;
+static volatile const __u8 *node_mac2 = mac_three;
+static volatile const __u8 *dst_mac2 = mac_four;
+
+#include "bpf_host.c"
+
+#include "lib/endpoint.h"
+#include "lib/ipcache.h"
+#include "lib/policy.h"
+
+#define FROM_NETDEV	0
+#define TO_NETDEV	1
+
+ASSIGN_CONFIG(bool, enable_extended_ip_protocols, true);
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PROG_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(max_entries, 2);
+	__array(values, int());
+} entry_call_map __section(".maps") = {
+	.values = {
+		[FROM_NETDEV] = &cil_from_netdev,
+		[TO_NETDEV] = &cil_to_netdev,
+	},
+};
+
+/* Send an IGMP packet from host to IGMP destination (allow all egress policy).
+ *
+ */
+PKTGEN("tc", "hostfw_igmp_egress")
+int hostfw_igmp_egress_pktgen(struct __ctx_buff *ctx)
+{
+	struct pktgen builder;
+	struct igmphdr *igmp;
+
+	/* Init packet builder */
+	pktgen__init(&builder, ctx);
+
+	igmp = pktgen__push_ipv4_igmp_packet(&builder,
+					     (__u8 *)node_mac, (__u8 *)dst_mac,
+					     NODE_IP, DST_IP,
+					     IGMP_HOST_MEMBERSHIP_REPORT);
+	if (!igmp)
+		return TEST_ERROR;
+
+	/* Calc lengths, set protocol fields and calc checksums */
+	pktgen__finish(&builder);
+
+	return 0;
+}
+
+SETUP("tc", "hostfw_igmp_egress")
+int hostfw_igmp_egress_setup(struct __ctx_buff *ctx)
+{
+	policy_add_egress_allow_all_entry();
+	endpoint_v4_add_entry(NODE_IP, 0, 0, ENDPOINT_F_HOST, HOST_ID,
+			      0, (__u8 *)node_mac, (__u8 *)node_mac);
+	ipcache_v4_add_entry(NODE_IP, 0, HOST_ID, 0, 0);
+	set_identity_mark(ctx, 0, MARK_MAGIC_HOST);
+
+	/* Jump into the entrypoint */
+	tail_call_static(ctx, entry_call_map, TO_NETDEV);
+	/* Fail if we didn't jump */
+	return TEST_ERROR;
+}
+
+CHECK("tc", "hostfw_igmp_egress")
+int hostfw_igmp_egress_check(const struct __ctx_buff *ctx)
+{
+	void *data, *data_end;
+	__u32 *status_code;
+
+	test_init();
+
+	data = (void *)(long)ctx_data(ctx);
+	data_end = (void *)(long)ctx->data_end;
+
+	if (data + sizeof(__u32) > data_end)
+		test_fatal("status code out of bounds");
+
+	status_code = data;
+
+	assert(*status_code == CTX_ACT_OK);
+
+	/* Check for egress CT entry */
+	struct ipv4_ct_tuple tuple = {
+		.daddr   = NODE_IP,
+		.saddr   = DST_IP,
+		.dport   = 0,
+		.sport   = 0,
+		.nexthdr = IPPROTO_IGMP,
+		.flags = TUPLE_F_OUT,
+	};
+	struct ct_entry *ct_entry = map_lookup_elem(get_ct_map4(&tuple), &tuple);
+
+	if (!ct_entry)
+		test_fatal("no CT entry found");
+
+	assert(ct_entry->packets == 1);
+
+	policy_delete_egress_entry();
+
+	test_finish();
+}
+
+/* Send an IGMP packet from the IGMP destination to host. Packet is allowed based on
+ * conntrack entry (no ingress policy).
```

Comment:
Aren't the different tests supposed to stay independent? As in, the second should be able to run on its own.

Reply:
I'm not aware of such a requirement. We seem to have a few instances of dependent tests - https://github.com/cilium/cilium/blob/main/bpf/tests/hostfw_host_iptables.c and https://github.com/cilium/cilium/blob/main/bpf/tests/inter_cluster_snat_clusterip_backend_lxc.c.